### ISSUE TITLE
How to set the same hotkey to multiple apps?

### ISSUE THREAD

**Issue by tipotto** [autokey/autokey#642]
## Classification:
UI/Usability

## Reproducibility:
Always

## AutoKey version:
0.95.10

## Used GUI:
Gtk

## Installed via:
package manager

## Linux distribution:
Kali linux v2021.3.0

## Summary:
I'd like to set the same hotkey that triggers a script to multiple apps, but couldn't figure out how to do it.

## Steps to reproduce:

1. Created two folders for browser and terminal
2. Set window filters for each directory.
3. Located the following script triggered with Ctrl+P into each directory.

up.py
```
keyboard.send_keys("<up>")
```

## Expected result:
Set the same hotkey that triggers a script to multiple apps.

## Actual result:
The following error occurred.
```
The hotkey '<ctrl+p>' is already in use by the script 'up'
Your changes have not been saved.
```

## Notes:
Thanks a lot for this useful tool. I also attempted the following method to apply the hotkey globally to multiple apps. However, it didn't work as expected. Appreciate if you give me some suggestions.

1. Created a folder with no window filter set.
2. Located up.py into the directory.
3. Set a hotkey (Ctrl+P) to the script.

**Comment by Elliria** [comment]
This would be perfect for our [Google list](https://groups.google.com/forum/#!forum/autokey-users) or our [Gitter chat](https://gitter.im/autokey/autokey) where more people are likely to see it and respond to it.

But since we're already here, the solution will depend on exactly what you'd like the script to do. I'm not sure if you want to control two programs from one script or take a different action based on which of two programs you're interacting with.

## Example 1:
This script offers a few different ways to control two programs, depending on what you'd like to do with them:
```
# Open Firefox and then immediately open Vivaldi:
subprocess.Popen(["firefox"])
subprocess.Popen(["vivaldi"])

# Open Firefox and then open Vivaldi only when Firefox has been closed:
subprocess.call(["firefox"])
subprocess.call(["vivaldi"])

# Open Firefox and then open Vivaldi only when Firefox has been closed:
import subprocess
subprocess.run(["firefox"])
subprocess.run(["vivaldi"])
```
Firefox and Vivaldi were used in the examples, but you could substitute any path or executable name for either or both of those and/or add other programs and/or add other commands before or after those.

## Example 2:
This script checks the class of the focused window to find out what's running and executes a custom action based on the result:
```
# Get the class of the currently-focused window:
winClass = window.get_active_class()

# Determine which action to take based on which window has the current focus:
if "Firefox" in winClass:
    # Zoom the contents of the window:
    output = "<ctrl>++"
elif "Chromium" in winClass:
    # Search the internet for hello world:
    output = "hello world+<enter>"
elif "Akregator" in winClass:
    # Move to the next unread item:
    output = "="
else:
    # Do nothing:
    output = ""

keyboard.send_keys(output)
```
You can change the Firefox, Chromium, and Akregator class names to any that you like, remove or add some, etc., and also change the actions that are taken depending on which window has the focus.

## Last, but not least:
This [Stack Overflow page](https://stackoverflow.com/questions/40697583/whats-the-difference-between-pythons-subprocess-call-and-subprocess-run) goes over how to choose between ```subprocess.call``` and ```subprocess.Popen``` and ```subprocess.run```.

**Comment by josephj11** [comment]
As @Elliria says, we'd need to know more specifically what you're trying to do - maybe with some pseudocode to describe it.

I have two scripts that both use the same hotkey. They have to have window filters (regexes) that are mutually exclusive for AutoKey to let you define them and work reliably. Mine are both triggered by Ctrl+P, but each runs a different script to automate a specific style of an application Print dialog.

However, it turns out that determining algorithmically whether two regexes are mutually exclusive is a difficult problem, so AutoKey just checks that they're not identical. If they are not actually mutually exclusive, then results are undefined, but AutoKey probably will use the first one it finds that matches.

**Comment by tipotto** [comment]
@Elliria 

Sorry for the late reply.
Appreciate for your detailed suggestion.
I modified my script based on example 2 and everything worked as expected.

The modified script is as follows:

```
winClass = window.get_active_class()

if "Vivaldi" in winClass or "qterminal" in winClass:
    keyboard.send_keys("<up>")
```

**Comment by tipotto** [comment]
@josephj11 

Sorry for the late reply.
What I wanted to do is trigger a common script with a specific key binding on multiple apps, and it worked perfectly by using ```  window.get_active_class ``` and if statement thanks to @Elliria's advice.

**Comment by tipotto** [comment]
There's one thing I have something to ask.
I suppose autokey doesn't send any keys on Emacs when the following script bound for Ctrl+G is executed. However, I'd like to trigger predefined Emacs function with Ctrl+G without making autokey affect on Emacs.

escape.py
Window filter is not set in this case.
```
if not "Emacs" in window.get_active_class():
    keyboard.send_keys("<escape>")
```

It's also the same for the case the following regex is set in window filter and the below script is run. (The below regex matches both window name and class. If not, escape key is sent on Emacs.)
```
^((?![Ee]macs).)*$
```

escape.py
```
keyboard.send_keys("<escape>")
```

Is there any way to exclude Emacs from the scope of autokey and work it out?

The reason why I'd like to do it is because it's not obvious which key is triggered inside Emacs in this case, so it's difficult to explicitly define keys to send. I created the script below to deal with it, but it worked differently from when autokey was off.

```
if "Emacs" in window.get_active_class():
    keyboard.send_keys("<ctrl>+g")
else:
    keyboard.send_keys("<escape>")
```

**Comment by Elliria** [comment]
I'm not sure I fully understand all of what you're trying to do and I have no knowledge of Emacs, but if you'd like AutoKey to ignore Emacs windows, this would be my "bull in a China shop" approach to it:
```python
# Get the class of the currently-focused window:
winClass = window.get_active_class()

# if this is an Emacs window:
if "Emacs" in winClass:
    # do nothing:
    pass
# if this is any other window:
else:
    # press the Esc key:
    keyboard.send_keys("<escape>")
```

**Comment by tipotto** [comment]
Thank you for your reply.
I attempted for pass statement, and no keys were sent to Emacs unfortunately.

What I want to do when pressing Ctrl+G is
1. Send escape key to all apps but Emacs
2. Make autokey not affect Emacs, and use Emacs the same way as when autokey is turned off.

As for Emacs, the behavior is different between two following cases. 
1. When Ctrl+G key is sent through autokey with ``` keyboard.send_keys("<ctrl>+g") ```.
2. When Ctrl+G is typed on Emacs with autokey turned off.

Therefore, I'd like autokey not to affect Emacs at all (I'd like to separate Emacs from other apps that autokey affects)

However, I assume maybe autokey affects all apps when activated and does not send any keys to the following apps.

- Apps excluded by Window filter.
- Apps you don't define the key to be sent explicitly in the if statement.

Sorry if my explanation is not clear.
Hope you understand what I want to do.

**Comment by Elliria** [comment]
I just reread this issue report and it seems that I missed your original issue entirely. In looking that over again, the error message you received was telling you that the hotkey combination you had chosen was already in use by the "up" script. That could be solved by using a different hotkey combination for this script or changing or removing the hotkey combination from the other script so that you can use it with this script. Hopefully that will solve the reported issue.

Meanwhile, the Emacs control is a bit tricky. The **pass** statement that I had suggested is a null statement and intentionally does nothing at all. I had thought that was what you wanted, so that's entirely my fault.

If I'm understanding what you're wanting the script to do, you'd like the **Ctrl+g** hotkey to be sent to Emacs and for Emacs to then use it the way it normally would if you had typed it. Is that right? I also get the impression that if any other window that isn't an Emacs window has the focus, you'd like the hotkey to **not** be sent to that window and the **Esc** key to be sent to the window instead. Is that right? Either or both of those should be possible, but it's vital to know that AutoKey always consumes the hotkey, so it's not sent to any window by default.

It should be as simple as this:
```python
if "Emacs" in window.get_active_class():
    keyboard.send_keys("<ctrl>+g")
else:
    keyboard.send_keys("<escape>")
```
But you mentioned that when AutoKey sends ```keyboard.send_keys("<ctrl>+g")``` to Emacs, it doesn't do the same thing as when you type **Ctrl+g** in Emacs. This may mean that it's being intercepted by Emacs or by your system or that Emacs wants to receive it in a slightly different way. The use of the plus key in that AutoKey command makes it so that the **Ctrl** key and the **g** key are pressed simultaneously. Perhaps you could try it without the plus to see whether it's accepted. That will press the **Ctrl** key and then press the **g** key, one after another. Perhaps that's what Emacs wants:
```python
if "Emacs" in window.get_active_class():
    keyboard.send_keys("<ctrl>g")
else:
    keyboard.send_keys("<escape>")
```
If that doesn't work, this might work. It uses the ```press_key()``` and ```release_key()``` methods to press and release the keys and should be interpreted by Emacs as though the keys were pressed simultaneously:
```python
if "Emacs" in window.get_active_class():
    keyboard.press_key("<ctrl>")
    keyboard.press_key("g")
    keyboard.release_key("g")
    keyboard.release_key("<ctrl>")
else:
    keyboard.send_keys("<escape>")
```
Last, but not least, I see that you had used a regex to exclude Emacs windows, but were wondering if there was a way to exclude Emacs from the scope and I figured an **if not** statement was worthy of a quick mention here, although AutoKey would still consume the hotkey with this one:
```python
if not "Emacs" in window.get_active_class():
    keyboard.send_keys("<escape>")
```

**Comment by Elliria** [comment]
This page may also be useful: [https://emacs.stackexchange.com/questions/39027/how-to-control-emacs-externally](https://emacs.stackexchange.com/questions/39027/how-to-control-emacs-externally)

**Comment by josephj11** [comment]
OK. A few comments.

- Your current approach of interrogating the window class in you script works. It does have the downsides that it consumes the hotkey so it won't get passed along to any application that your filter matches and you can't have other scripts that use the same hotkey.
- This is where it gets interesting. The "right" way to do this _would_ be to use window filters in AutoKey. This way the hotkey can be reused for other scripts with mutually exclusive window filters and be passed as-is to any window that doesn't match the filter.

_But_ there's a deficiency in AutoKey that makes most negative window filters fail. I believe there's code to fix it, but it didn't make it into our current release or into our current beta release. See #161 and #305 for all the gory details.

As it stands now, AutoKey compares the window filter to _both_ the window title _and_ to the window class and succeeds if _either_ one matches. This necessitates making a complex filter regex that will fail to match both the title and the class of the excluded window, but still matches the desired windows.

This will be fixed, hopefully, soon after 0.96.0 is released. There will be a whole new filtering system which will be much more powerful and easier to use.

- For now, you can keep your script and filters as is, but have your script start by checking the window class and/or title for emacs. If the test succeeds, emit a Ctrl+G and exit.
